agent/xds/listeners_ingress.go

```diff
@@ -154,21 +137,104 @@ func (s *ResourceGenerator) makeIngressGatewayListeners(address string, cfgSnap
+func makeDownstreamTLSContextFromSnapshotListenerConfig(cfgSnap *proxycfg.ConfigSnapshot, listenerCfg structs.IngressListener) (*envoy_tls_v3.DownstreamTlsContext, error) {
+	var downstreamContext *envoy_tls_v3.DownstreamTlsContext
+
+	tlsContext, err := makeCommonTLSContextFromSnapshotListenerConfig(cfgSnap, listenerCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	if tlsContext != nil {
+		downstreamContext = &envoy_tls_v3.DownstreamTlsContext{
+			CommonTlsContext:         tlsContext,
+			RequireClientCertificate: &wrappers.BoolValue{Value: false},
+		}
+	}
+
+	return downstreamContext, nil
+}
+
+func makeCommonTLSContextFromSnapshotListenerConfig(cfgSnap *proxycfg.ConfigSnapshot, listenerCfg structs.IngressListener) (*envoy_tls_v3.CommonTlsContext, error) {
+	var tlsContext *envoy_tls_v3.CommonTlsContext
+
+	// Enable connect TLS if it is enabled at the Gateway or specific listener
+	// level.
+	gatewayTLSCfg := cfgSnap.IngressGateway.TLSConfig
+
+	// TODO: should it be possible to explicitly _disable_ TLS on a listener if it's
+	// enabled on the gateway?
+	connectTLSEnabled := gatewayTLSCfg.Enabled ||
+		(listenerCfg.TLS != nil && listenerCfg.TLS.Enabled)
+
+	tlsCfg, err := resolveListenerTLSConfig(&gatewayTLSCfg, listenerCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	if tlsCfg.SDS != nil {
+		// Set up listener TLS from SDS
+		tlsContext = makeCommonTLSContextFromGatewayTLSConfig(*tlsCfg)
+	} else if connectTLSEnabled {
+		tlsContext = makeCommonTLSContextFromLeaf(cfgSnap, cfgSnap.Leaf(), makeTLSParametersFromGatewayTLSConfig(*tlsCfg))
+	}
+
+	return tlsContext, nil
+}
+
+func resolveListenerTLSConfig(gatewayTLSCfg *structs.GatewayTLSConfig, listenerCfg structs.IngressListener) (*structs.GatewayTLSConfig, error) {
+	var mergedCfg structs.GatewayTLSConfig
+
+	resolvedSDSCfg, err := resolveListenerSDSConfig(gatewayTLSCfg.SDS, listenerCfg.TLS, listenerCfg.Port)
+	if err != nil {
+		return nil, err
+	}
+
+	mergedCfg.SDS = resolvedSDSCfg
+
+	if gatewayTLSCfg != nil {
+		mergedCfg.TLSMinVersion = gatewayTLSCfg.TLSMinVersion
+		mergedCfg.TLSMaxVersion = gatewayTLSCfg.TLSMaxVersion
+		mergedCfg.CipherSuites = gatewayTLSCfg.CipherSuites
+	}
+
+	if listenerCfg.TLS != nil {
+		if listenerCfg.TLS.TLSMinVersion != types.TLSVersionUnspecified {
+			mergedCfg.TLSMinVersion = listenerCfg.TLS.TLSMinVersion
+		}
+		if listenerCfg.TLS.TLSMaxVersion != types.TLSVersionUnspecified {
+			mergedCfg.TLSMaxVersion = listenerCfg.TLS.TLSMaxVersion
+		}
+		if len(listenerCfg.TLS.CipherSuites) != 0 {
+			mergedCfg.CipherSuites = listenerCfg.TLS.CipherSuites
+		}
+	}
+
+	// Validate. Configuring cipher suites is only applicable to connections negotiated
+	// via TLS 1.2 or earlier. Other cases shouldn't be possible as we validate them at
+	// input but be resilient to bugs later.
+	switch {
+	case mergedCfg.TLSMinVersion >= types.TLSv1_3 && len(mergedCfg.CipherSuites) != 0:
```

Comment:
TODO: This needs to change from a relative order/comparison check to equality now that TLSVersion is a `string` type instead of `int`.

Reply:
😱 this still compiles for now somehow, I'm assuming it's comparing the length of the strings??

Reply:
I _think_ this additional validation is still worthwhile if the _gateway_ defines `TLSMinVersion = "TLSv1_3"` and a _listener_ attempts to set `CipherSuites`, which wouldn't be caught in the config entry validation until the TLS configs are merged here - however that would be quite an odd edge case.

Reply:
This now checks if mergedCfg.TLSMinVersion is in the set of TLS versions with configurable cipher suites (TLS 1.0, 1.1, 1.2 and "auto" which currently defaults to 1.0 in Envoy).